File: 214/projects/05/java/Roots.java
/* Roots.java is a method and driver for finding the roots of a 
 * quadratic equation using the quadratic formula.
 * 
 * By: Duncan Van Keulen on 3-6-2020
 ***************************************************************/

import java.util.Scanner;
import java.lang.Math;

public class Roots {
    public static void main(String[] args) {
        Scanner keyboard = new Scanner(System.in);

        // Get inputs for a, b, c
        System.out.print("Quadratic Root Finder\nEnter a: ");
        double a = keyboard.nextDouble();

        System.out.print("Enter b: ");
        double b = keyboard.nextDouble();

        System.out.print("Enter c: ");
        double c = keyboard.nextDouble();

        double[] rootsArr = new double[2];

        // Only display roots if they are valid
        if (root(a, b, c, rootsArr)) {
            System.out.println("The roots are " + rootsArr[0] + " and " + rootsArr[1]);
        }
    }

    /*
     * root() calculates the quadratic roots of an equation using the quadratic
     * formula will return true if calculation succeeded and false if it failed.
     * double a, b, c are the a, b, c numbers in the quadratic formula double[]
     * rootsArr is the "pass by reference" array that will contain the roots
     *******************************************************************************/
    public static boolean root(double a, double b, double c, double[] rootsArr) {
        if (a != 0) {
            double arg = Math.pow(b, 2.0) - 4 * a * c;
            if (arg >= 0) {
                rootsArr[0] = (-b + Math.sqrt(arg)) / (2 * a);
                rootsArr[1] = (-b - Math.sqrt(arg)) / (2 * a);
                return true;
            } else {
                System.out.println("root(): b^2 - 4ac is negative!");
                rootsArr[0] = 0.0;
                rootsArr[1] = 0.0;
                return false;
            }
        } else {
            System.out.println("root(): a is zero!");
            rootsArr[0] = 0.0;
            rootsArr[1] = 0.0;
            return false;
        }
    }
}

File: code/projects/05/clojure/src/roots.clj
;;;; roots.clj  string into two substrings.
;;;;
;;;; Input: a, b, c, corresponding values to the quadratic function values
;;;; Output: the roots of the quadratic function
;;;;
;;;; Usage: clojure -m roots
;;;;
;;;; By: Duncan Van Keulen
;;;; Date: 2-9-2020
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(ns roots)      ; the name of the program

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; roots() calculates the roots of a quadratic function.  
;;; Receive: a, b, c: values for quadratic equation.
;;; PRE: a != 0, b^2 - 4ac is positive.
;;; Return: the roots of the function
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn roots [a b c]
  (if (= a 0) 
    ;; a is 0
    (println "roots: a is zero!")
  
    ;; else: a is not 0
    (let
      ;; compute the discriminant
      [arg (- (* b b) (* 4.0 (* a c)))]
      ;; make sure discriminant is not negative
      (if (>= arg 0.0) 
        (do
          ;; calculate the roots
          (def root1 (/ (+ (- 0.0 b) (Math/sqrt arg)) (* 2.0 a)))
          (def root2 (/ (- (- 0.0 b) (Math/sqrt arg)) (* 2.0 a)))
          
          ;; return a vector with the answers a bool indicating it suceeded
          (vector root1 root2 true)
        )
        ;; else: discriminant is negative
        (do 
          (println "roots: discriminant is negative!")
        )
      )
    )
  )
)


;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; main function
;;; Input: a, b, c the values in the quadratic function
;;; Output: the roots of that quadratic function
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn -main []
  (print "To calculate the roots of a quadratic equation,\nEnter a: ") (flush)
  (let
    ;; get the value of a
    [a (read)]
    (print "Enter b: ") (flush)
    (let 
      ;; get the value of b
      [b (read)]
    (print "Enter c: ") (flush)
      (let
        ;; get the value of c
        [c (read)]

        ;; get the result and grab each individual thing returned from the vector
        (let [
          result (roots a b c)
          root1 (get result 0)
          root2 (get result 1)
          succeeded (get result 2)
          ]
          ;; only print if it succeeded
          (if succeeded 
            (do 
              (println (format "The roots are %f and %f" root1 root2))
            )
          )
          ; else nothing
        )
      )
    )
  )
)


